src/ir/access_path_root.cc

```diff
@@ -0,0 +1,34 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#include "src/ir/access_path_root.h"
+
+#include "src/ir/datalog_print_context.h"
+
+// The classes in this file describe the root of an AccessPath. At the moment,
+// we have only two types of roots: a HandleConnectionSpecAccessPathRoot and
+// a HandleConnectionAccessPathRoot, describing a HandleConnectionSpec in a
+// ParticleSpec and a Handle in a Particle, respectively.
+namespace raksha::ir {
+
+std::string AccessPathRoot::ToDatalog(const DatalogPrintContext &ctxt) const {
+  const auto *instantiation_map = ctxt.instantiation_map();
+  if (instantiation_map == nullptr) return this->ToString();
+  auto find_res = instantiation_map->find(*this);
+  return (find_res != instantiation_map->end()) ? find_res->second.ToString()
+                                                : this->ToString();
```

Comment:
Also, this seems somewhat overly permissive. Within a `Particle`, we currently expect to find a match for every `AccessPath` present. This will not always be the case (for instance, stores will not be affected by the instantiation map) , but we don't currently handle the cases that would be exceptions. For now, I think it might be more appropriate to `CHECK` that the result is not `end()`.

Reply:
See comment above. In the case you are referring to, the problem will be caught here:
https://github.com/google-research/raksha/blob/a24d1baeac7f6c107f6d7321d9e373369c1527fa/src/ir/access_path_root.h#L45